Rewrote platform architecture docs (#371)
chalk-ai/docs: architecture.mdx

---

Chalk’s online query serving platform is architected to

- retrieve fresh features across heterogeneous sources with the minimum possible latency
- orchestrate and execute complex transformations on structured and unstructured data
- uphold enterprise-grade guarantees for security, observability, and reliability

In short, Chalk was built to get the right data from the right place at the right time.

> ## Onlnine serving architecture

Reviewer: ```suggestion
---

## Onlnine serving architecture
```

Reviewer: (typo in online and missing a line break)

Let's examine how the pieces work together to compute and serve **[features](/docs/features)**.

Suppose that you want to compute a set of features for making a decision about a request from a user:

1. Your application sends an [HTTP request](/docs/query-basics) to Chalk's serving API for features
2. Chalk's query planner generates an optimized execution plan by analyzing feature dependencies and available data sources.
3. Chalk's compute engine
    - retrieves fresh values from underlying sources with [Resolvers](/docs/resolver-overview)
    - pulls from Chalk’s low-latency online storage e.g. tiled / materialized features or [cached features](/docs/reverse-etl)
    - executes the generated query plan
4. Returns computed features in response
5. Newly computed features values are logged for reuse and auditability

![Chalk architecture diagram](/img/architecture-diagram.png)

> This entire pipeline—from SQL queries and API calls to response—runs in less than 10ms, even with heterogeneous data sources and complex logic.

Reviewer: ```suggestion
This entire pipeline—from SQL queries and API calls to response—runs in less than 5ms, even with heterogeneous data sources and complex logic.
```

Chalk uses many techniques to reduce latency, such as:

- Automatic parallelism across pipeline stages
- Vectorization of pipeline stages that are written using scalar syntax
- Statistics-informed join planning
- Low-latency key/value stores (like [Redis](https://redis.com/), [BigTable](https://cloud.google.com/bigtable), or [DynamoDB](https://aws.amazon.com/dynamodb/))
- Transpilation of Python code into native code with Chalk's [symbolic Python interpreter](https://chalk.ai/blog/symbolic-python-interpreter)

### Data orchestration

Chalk eliminates the complexity of orchestrating data and ETL pipelines by building a dependency graph (DAG) of your features — which are defined using Python.

> At inference time, Chalk dynamically--without manual configuration--builds query plans (sub-graphs of your feature DAG) based on the features you request.

Reviewer: ```suggestion
At inference time, Chalk dynamically builds query plans (subgraphs of your feature DAG) without manual configuration, based on the features you request.
```

> Write feature definitions in Python and Chalk automatically

Reviewer: ```suggestion
Write feature definitions in Python, and Chalk automatically
```

With Chalk, data teams get

- Comprehensive logging & debugging
    - Resolver execution details
    - Feature computations
    - Cache misses
    - Data source failures
    - Feature staleness
    - Feature ownership tracking
- Metrics and performance monitoring
    - Features: request volumes, computation times, and error rates
    - Models: inference latency, prediction accuracy, and resource utilization
    - System: query throughput, system latency, and pipeline health
- End-to-end request tracing
    - Track the inputs and outputs of every step in any run
    - Lineage tracking from data source to final predictions

Easily, build your own views and set up custom dashboards to visualize your metrics and configure smart alerts with custom formulas that notify you instantly when thresholds are crossed or anomalies are detected.

![Metrics and chart creation](/img/architecture-monitoring-chart_creation.png)

This flexibility to configure and define your own metrics makes it easy to answer common questions such as how often certain features are computed, how long individual computations take, and what the average value for a feature is.

> ## Architecting maintainable and scalable systems for enterprise success

Reviewer: ```suggestion
---

## Architecting maintainable and scalable systems for enterprise success
```
